### PR TITLE
#1404 Serial Monitor message history

### LINKED ISSUE #1404: Serial Monitor message history
### Describe the request

In Arduino IDE 1.x, you can 'up arrow' through the recently sent commands, to send the same one a second time, or even go further back in history.

### Describe the current behavior

none, this functionality is missing.

### Arduino IDE version

IDE 2.0.0-rc9.3

### Operating system

macOS

### Operating system version

Monterey

### Additional context

_No response_

### Issue checklist

- [X] I searched for previous requests in [the issue tracker](https://github.com/arduino/arduino-ide/issues?q=)
- [X] I verified the feature was still missing when using the latest [nightly build](https://github.com/arduino/arduino-ide#nightly-builds)
- [X] My request contains all necessary details

### PR DESCRIPTION
**Pull request by dwightfowler-cd**
### Motivation
Close #1404

### Change description
Implement history for Serial Monitor input using up-arrow / down-arrow.

### Other information
N/A

### Reviewer checklist

* [ ] PR addresses a single concern.
* [ ] The PR has no duplicates (please search among the [Pull Requests](https://github.com/arduino/arduino-ide/pulls) before creating one)
* [ ] PR title and description are properly filled.
* [ ] Docs have been added / updated (for bug fixes / features)

### CODE REVIEW

**Comment by dwightfowler-cd** [comment]
My apologies for commit [26dc4c5](https://github.com/arduino/arduino-ide/pull/1453/commits/26dc4c5972bb9127d0bd5fbf64b29013905f4020) and [7851637](https://github.com/arduino/arduino-ide/pull/1453/commits/7851637c33cf9981558a04e7478ea7faed510d76). I forgot that PRs into my personal `main` branch get added over here at `arduino/arduino-ide`.

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx [hunk 1]
```diff
@@ -6,6 +6,75 @@ import { BoardsServiceProvider } from '../../boards/boards-service-provider';
...
+  public Init(size: number = 100)
+  {
+    this.ring = [];
+    this.size = (size > 0) ? size : 1;
+    this.begin = 0;
+    this.index = 0;
+    this.end = -1;
+  }
+
+  public Push(val: string): number {
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:33]: Method names start with lowercase and have camel case styling. Please change to `init`, `prev`, etc.

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx [hunk 2]
```diff
@@ -6,6 +6,75 @@ import { BoardsServiceProvider } from '../../boards/boards-service-provider';
...
+
+  constructor(size: number = 100) {
+    this.Init = this.Init.bind(this);
+    this.Push = this.Push.bind(this);
+    this.Prev = this.Prev.bind(this);
+    this.Next = this.Next.bind(this);
+    this.Init(size);
+  }
+
+  public Init(size: number = 100)
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:24]: You `Init` the `RingList` from the `constructor` function; please change the visibility to `private`.

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx [hunk 3]
```diff
@@ -6,6 +6,75 @@ import { BoardsServiceProvider } from '../../boards/boards-service-provider';
...
+    }
+    this.ring[this.end] = val;
+    this.index = this.end;
+
+    return this.index;
+  }
+
+  public Prev(): string {
+    if (this.ring.length < 1) {
+      return "";
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:54]: Please use single-quotes for strings. `'` is OK, `"` is only OK if you want to escape `'`.

Examples:
`'OK'`, ✅ 
`"NOK"`, 🔥 
`"If you escape single quotes: It's OK."`. ✅

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx [hunk 4]
```diff
@@ -110,7 +180,17 @@ export class SerialMonitorSendInput extends React.Component<
     if (keyCode) {
       const { key } = keyCode;
       if (key === Key.ENTER) {
+        // NOTE: order of operations is critical here. Push the current state.text
+        // onto the history stack before sending. After sending, state.text is empty
+        // and you'd end up pushing '' onto the history stack.
+        if (this.state.text.length > 0) this.state.history.Push(this.state.text);
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:186]: Please format + indent your code:

Bad:
```ts
if (this.state.text.length > 0) this.state.history.Push(this.state.text);
```

Good:
```ts
if (this.state.text.length > 0) {
  this.state.history.Push(this.state.text);
}
```

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx [hunk 5]
```diff
@@ -110,7 +180,17 @@ export class SerialMonitorSendInput extends React.Component<
...
         this.onSend();
+      } else
+      if (key === Key.ARROW_UP) {
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:189]: This should be:

```ts
if (condition) {
  // code
} else if (other_condition) {
  // code
}
```

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx [hunk 6]
```diff
@@ -6,6 +6,75 @@ import { BoardsServiceProvider } from '../../boards/boards-service-provider';
 import { MonitorModel } from '../../monitor-model';
 import { Unknown } from '../../../common/nls';
 
+class RingList {
+  protected ring: string[];
+  protected size: number;
+  protected begin: number;
+  protected index: number;
+  protected end: number;
```
Comment (kittaakos) [inline arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:14]: Should be `private`.

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx (hunk 1 above)
Comment (kittaakos) [reply arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:33]: The `public` keywords are redundant. Please remove them. It should be:

```ts
push(val: string): number {
  // code
}

prev(): string {
  // code
}
```

**Review by kittaakos — COMMENTED** [review]
Thank you for your contribution; I added a few remarks and am happy to try out your changes after your updates.

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx (hunk 2 above)
Comment (dwightfowler-cd) [reply arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:24]: Done

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx (hunk 1 above)
Comment (dwightfowler-cd) [reply arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:33]: Done

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx (hunk 3 above)
Comment (dwightfowler-cd) [reply arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:54]: Done

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx (hunk 4 above)
Comment (dwightfowler-cd) [reply arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:186]: Done

File: arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx (hunk 6 above)
Comment (dwightfowler-cd) [reply arduino-ide-extension/src/browser/serial/monitor/serial-monitor-send-input.tsx:14]: Done

**Comment by kittaakos** [comment]
@dwightfowler-cd, thank you for the updates. You have [opened a new PR](https://github.com/arduino/arduino-ide/pull/1467) instead of updating this one. Which one do you want me to review? Can you create one PR instead of opening a new one for each commit?

**Comment by kittaakos** [comment]
Closing in favor of #1467.

**Pull request closed without merging**